parla/task_runtime.py

```diff
@@ -349,118 +347,163 @@ def result(self):
+        assert isinstance(self.req, EnvironmentRequirements), \
+            "Task was not assigned a specific environment requirement before running."
+        try:
+            # A default state to avoid exceptions during catch
+            task_state = TaskException(RuntimeError("Unknown fatal error"))
+            # Run the task and assign the new task state
+            try:
+                assert isinstance(self._state, TaskRunning)
+                # TODO(lhc): This assumes Parla only has two devices.
+                #            The reason why I am trying to do is importing
+                #            Parla's cuda.py is expensive.
+                #            Whenever we import cuda.py, cupy compilation
+                #            is invoked. We should remove or avoid that.
+
+                # First, create device/stream/event instances.
+                # Second, gets the created event instance.
+                # Third, it scatters the event to successors who wait for
+                # the current task.
+                env = self.req.environment
+                with _scheduler_locals._environment_scope(env), env:
+                    events = env.get_events_from_components()
+                    self._wait_for_predecessor_events(env)
+                    task_state = self._run()
+                    # Events could be multiple for multiple devices task.
+                    env.record_events()
+                    if len(events) > 0:
+                        # If any event created by the current task exist,
+                        # notify successors and make them wait for that event,
+                        # not Parla task completion.
+                        self._notify_successors_mutex(events)
+                    env.sync_events()
+                task_state = task_state or TaskCompleted(None)
+            except Exception as e:
+                task_state = TaskException(e)
+                logger.exception("Exception in task")
+            finally:
+                logger.info("Finally for task %r", self)
+
+                ctx = get_scheduler_context()
+                self._post_run(ctx)
+
+                # Protect the case that it notifies successors and
+                # any successor task is spawned before setting state.
+                with self._mutex:
+                    # Regardless of the previous notification,
+                    # (So, before leaving the current run(), the above)
+                    # it should notify successors since
+                    # new successors could be added after the above
+                    # notifications, while other devices are running
+                    # their kernels asynchronously.
+                    self._notify_successors()
+                    self._set_state(task_state)
+
+        except Exception as e:
+            logger.exception("Task %r: Exception in task handling", self)
+            raise e
+
+
```

Comment:
Perhaps some portions of `run` could be further encapsulated into their own functions to make this cleaner

Reply:
This PR is already big due to long delay. Let's put some finer improvement in later PR.